### DISCUSSION TITLE
Handling missed messages for subscriptions incase of websocket reconnections.

### DISCUSSION THREAD

**Discussion by 97amarnathk** [urql-graphql/urql#1598] [Q&A] [answered]
Hey, 

I have the following urql client setup:

* use graphql-ws for subscription handling
* use urql on react

For realtime messaging on my app, I am using urql in the following way:

1. Fetch the initial messages over a query. 
2. For any upcoming messages, create a subscription 

This works fine in the happy case that the underlying websocket is connected.

However during socket reconnects (which can be due to auth token expiry), It may take a few seconds for the socket and subsciption client to reconnect. In this duration , a lot of messages might have arrived in the backend, but werent delivered to me. This might lead to some messages not appearing in the frontend

A way to fix this might be to do the folowing:
1. on a disconnect, the subscriptionClient retries a connection. 
2. As soon as the client is connected, fetch a query similar to the subscription query , to get the missed messages.
3. while query is being fetched, load all subscription messages in a queue (so that query results are added first, and then subscription results, as they came later)

My doubt is how do I implement this in practice in urql?

**Comment by kitten** [comment, marked as answer]
There's nothing that we can do about that per se. urql doesn't know about the transport protocol you're using and its implementation for subscriptions. So it also doesn't know that some subscriptions implementation aren't queue-based and will lose events, so ultimately it's up to you to define a system that ensures that queries and subscriptions don't lose events and refresh as needed 😅 unfortunately there simply isn't a system we could add that does this automatically.

You could do some trickery around what your subscription can deliver though, like an indicator of when it may be suitable to refresh the query based on a time stamp or reconnection
